Google.Api.Gax.Grpc.Tests/ApiCallTest.cs

```diff
@@ -132,8 +136,310 @@ public void WithGoogleRequestParam()
+        {
+            CallSettings syncCallSettings = null;
+            CallSettings asyncCallSettings = null;
+            var call0 = new ApiCall<ExtractedRequestParamRequest, SimpleResponse>(
+                (req, cs) => { asyncCallSettings = cs; return null; },
+                (req, cs) => { syncCallSettings = cs; return null; },
+                null);
+
+            // call corresponding to the following routing parameters:
+            // { field: "table_name", path_template: "{project_id=projects/*}/**" }
+            // { field: "table_name", path_template: "projects/*/{instance_id=instances/*}/**" }
+            var call1 = call0
+                .WithExtractedGoogleRequestParam(
+                    new HeaderParameterExtractor<ExtractedRequestParamRequest>("project_id",
+                        new Regex("^(?<project_id>projects/[^/]+)(?:/.*)?$"), request => request.TableName)
+                        .Add("instance_id",
+                            new Regex("^projects/[^/]+/(?<instance_id>instances/[^/]+)(?:/.*)?$"), request => request.TableName));
+            call1.Sync(new ExtractedRequestParamRequest { TableName = fieldValue}, null);
+            call1.Async(new ExtractedRequestParamRequest { TableName = fieldValue }, null);
+
+            CallSettingsTest.AssertRoutingHeader(syncCallSettings, expectedHeader);
+            CallSettingsTest.AssertRoutingHeader(asyncCallSettings, expectedHeader);
+        }
+
+        /// <summary>
+        /// Extracting multiple routing header key-value pairs by matching
+        /// several path templates on multiple request fields.
+        /// </summary>
+        [Theory]
+        [InlineData("projects/100/instances/200/tables/300", "profiles/profile_17", "project_id=projects/100&routing_id=profiles/profile_17")]
+        [InlineData("projects/100", "", "project_id=projects/100")]
+        [InlineData("projects/100", null, "project_id=projects/100")]
+        [InlineData(null, null, null)]
+        public void WithExtractedGoogleRequestParam_MultipleFields(string tableNameValue, string appProfileIdValue, string expectedHeader)
+        {
+            CallSettings syncCallSettings = null;
+            CallSettings asyncCallSettings = null;
+            var call0 = new ApiCall<ExtractedRequestParamRequest, SimpleResponse>(
+                (req, cs) => { asyncCallSettings = cs; return null; },
+                (req, cs) => { syncCallSettings = cs; return null; },
+                null);
+
+            // call corresponding to the following routing parameters:
+            // { field: "table_name", path_template: "{project_id=projects/*}/**" }
+            // { field: "app_profile_id", path_template: "{routing_id=**}" }
+            var call1 = call0
+                .WithExtractedGoogleRequestParam(
+                    new HeaderParameterExtractor<ExtractedRequestParamRequest>("project_id",
```

Comment:
If we had a parameterless constructor, and could accept strings instead of Regex (maybe just as a convenience) I think we could format these tests more simply (also simpler when just testing the extractor):

```csharp
var extractor = new HeaderParameterExtractor<ExtractedRequestParamRequest>()
    .WithParameter("project_id", "^(?<project_id>projects/[^/]+)(?:/.*)?$", request => request.TableName)
    .WithParameter("routing_id", "^(?<routing_id>.*)$", request => request.AppProfileId);
```

It could be even shorter (and less repetitive) if we either omitted the group name or used it as the parameter name...

Reply:
👍  done